x-pack/plugins/encrypted_saved_objects/index.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import crypto from 'crypto';
+import { Request } from 'hapi';
+import { Root } from 'joi';
+import { Legacy } from 'kibana';
+
+import { EncryptedSavedObjectsClientWrapper } from './server/lib/encrypted_saved_objects_client_wrapper';
+import {
+  EncryptedSavedObjectsService,
+  EncryptedSavedObjectTypeRegistration,
+} from './server/lib/encrypted_saved_objects_service';
+import { createLogger } from './server/lib/logger';
+
+export const encryptedSavedObjects = (kibana: any) =>
+  new kibana.Plugin({
+    id: 'encrypted_saved_objects',
+    configPrefix: 'xpack.encrypted_saved_objects',
+    require: ['kibana', 'elasticsearch', 'xpack_main'],
+
+    config(Joi: Root) {
+      return Joi.object({
+        enabled: Joi.boolean().default(true),
+        encryptionKey: Joi.string().min(32),
+        audit: Joi.object({ enabled: Joi.boolean().default(false) }),
+      }).default();
+    },
+
+    async init(server: Legacy.Server) {
+      const config = server.config();
+      let encryptionKey = config.get<string>('xpack.encrypted_saved_objects.encryptionKey');
+      if (encryptionKey == null) {
+        server.log(
+          ['warning', 'encrypted-saved-objects'],
+          'Generating a random key for xpack.encrypted_saved_objects.encryptionKey. To be able ' +
+            'to decrypt encrypted saved objects attributes after restart, please set ' +
+            'xpack.encrypted_saved_objects.encryptionKey in kibana.yml'
+        );
+
+        encryptionKey = crypto.randomBytes(16).toString('hex');
+      }
+
+      const service = Object.freeze(
```

Comment:
__note:__ I've added `TODO:` comment at the bottom of this file, but all these `Object.freeze` don't make this code more secure. If we don't freeze `EncryptedSavedObjectsService.prototype` and `EncryptedSavedObjectsClientWrapper.prototype` malicious plugin can replace any of the functions it wants, but at the same time plugin can just read config object with encryption key and do whatever it wants :)

I'm pretty sure we're not trying to protect ourselves from malicious plugins since there is no way we can do that properly in the near future anyway and it's admin's responsibility to audit 3rd-party plugins they enable.

Having said that, I still try to use read-only TS annotations where possible as it sounds like a good practice to me and sparingly use `Object.freeze` even though it doesn't bring much value in most of the cases (maybe I should stop doing that for the objects that don't leave plugin boundaries?).

What are your thoughts on that?

Reply:
Okay, then we're on the same page, thanks :+1: